acme/wrappers/open_spiel_wrapper.py

```diff
@@ -0,0 +1,145 @@
+# python3
+# Copyright 2018 DeepMind Technologies Limited. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Wraps an OpenSpiel RL environment to be used as a dm_env environment."""
+
+from typing import List, NamedTuple
+
+from acme import specs
+from acme import types
+import dm_env
+import numpy as np
+from open_spiel.python import rl_environment
+import pyspiel
+
+
+class OLT(NamedTuple):
+  """Container for (observation, legal_actions, terminal) tuples."""
+  observation: types.Nest
+  legal_actions: types.Nest
+  terminal: types.Nest
+
+
+# TODO Wrap the underlying OpenSpiel game directly instead of OpenSpiel's
+# rl_environment?
+class OpenSpielWrapper(dm_env.Environment):
+  """Environment wrapper for OpenSpiel RL environments."""
+
+  # Note: we don't inherit from base.EnvironmentWrapper because that class
+  # assumes that the wrapped environment is a dm_env.Environment.
+
+  def __init__(self, environment: rl_environment.Environment):
+    self._environment = environment
+    self._reset_next_step = True
+    assert environment._game.get_type(
+    ).dynamics == pyspiel.GameType.Dynamics.SEQUENTIAL, (
+        "Currently only supports sequential games.")
+
+  def reset(self) -> dm_env.TimeStep:
+    """Resets the episode."""
+    self._reset_next_step = False
+    open_spiel_timestep = self._environment.reset()
+    assert open_spiel_timestep.step_type == rl_environment.StepType.FIRST
+    observations = self._convert_observation(open_spiel_timestep)
+    return dm_env.restart(observations)
+
+  def step(self, action: types.NestedArray) -> dm_env.TimeStep:
+    """Steps the environment."""
+    if self._reset_next_step:
+      return self.reset()
+
+    open_spiel_timestep = self._environment.step(action)
+
+    if open_spiel_timestep.step_type == rl_environment.StepType.LAST:
+      self._reset_next_step = True
+
+    observations = self._convert_observation(open_spiel_timestep)
+    rewards = np.asarray(open_spiel_timestep.rewards)
+    discounts = np.asarray(open_spiel_timestep.discounts)
+    step_type = open_spiel_timestep.step_type
+
+    if step_type == rl_environment.StepType.FIRST:
+      step_type = dm_env.StepType.FIRST
+    elif step_type == rl_environment.StepType.MID:
+      step_type = dm_env.StepType.MID
+    elif step_type == rl_environment.StepType.LAST:
+      step_type = dm_env.StepType.LAST
+    else:
+      raise ValueError("Did not recognize OpenSpiel StepType")
```

Comment:
Print the type it didn't recognize for debugging.

Reply:
Fixed